File: apps/server/app/api/v1/discovery.py
from fastapi import APIRouter,Depends
from sqlalchemy.orm import Session
from app.core.database import get_db
from app.models.song import Song
from app.schemas.common import ApiResponse

router=APIRouter(prefix='/discovery',tags=['discovery'])

@router.get('/hot',response_model=ApiResponse[list])
def hot_songs(db:Session=Depends(get_db)):
    items=db.query(Song).filter(Song.deleted_at.is_(None),Song.is_public.is_(True)).order_by((Song.like_count+Song.favorite_count+Song.view_count).desc()).limit(10).all()
    return ApiResponse(data=[{'id':x.id,'title':x.title,'artist':x.artist_name,'likes':x.like_count} for x in items])

@router.get('/keywords',response_model=ApiResponse[list])
def hot_keywords():
    return ApiResponse(data=['晴天','成都','周杰伦','民谣','新手弹唱','AI原创'])

@router.get('/recommend',response_model=ApiResponse[list])
def recommend(db:Session=Depends(get_db)):
    items=db.query(Song).filter(Song.deleted_at.is_(None),Song.is_public.is_(True)).order_by(Song.practice_count.desc(),Song.created_at.desc()).limit(8).all()
    return ApiResponse(data=[{'id':x.id,'title':x.title,'artist':x.artist_name} for x in items])

File: apps/miniprogram/src/api/discovery.ts
import { request } from './provider'

export interface DiscoverySong {
  _id?: string
  id?: number | string
  title: string
  artist?: string
  artist_name?: string
  likes?: number
  like_count?: number
}

function normalizeSongs(items: any[] = []): DiscoverySong[] {
  return items.map((item) => ({
    ...item,
    id: item.id || item._id,
    artist: item.artist || item.artist_name || '谱友作品',
    likes: item.likes || item.like_count || 0,
  }))
}

export async function getHotSongs() {
  const result = await request('discovery', { action: 'hot' })
  const data = result?.data || result || []
  return normalizeSongs(Array.isArray(data) ? data : data.hot || [])
}

export async function getHotKeywords() {
  const result = await request('discovery', { action: 'keywords' })
  return result?.data || result || []
}

export async function getRecommendSongs() {
  const result = await request('discovery', { action: 'recommend' })
  const data = result?.data || result || []
  return normalizeSongs(Array.isArray(data) ? data : data.recommend || [])
}

export async function getHomeDiscovery() {
  const result = await request('discovery', { action: 'home' })
  const data = result?.data || result || {}
  return {
    keywords: data.keywords || [],
    hot: normalizeSongs(data.hot || []),
    recommend: normalizeSongs(data.recommend || []),
  }
}
